src/libcrun/cgroup.c

```diff
@@ -690,8 +692,108 @@ systemd_job_removed (sd_bus_message *m, void *userdata, sd_bus_error *error arg_
   return 0;
 }
 
+/* Parse a gvariant string.  Support only a subset of types, just enough for systemd .  */
+static int
+append_systemd_annotation (sd_bus_message *m,
+                           const char *name,
+                           const char *value,
+                           libcrun_error_t *err)
+{
+  const char *it;
+  int sd_err;
+
+  while (*value == ' ')
+    value++;
+
+  it = value;
+
+  if (*it == '\'')
+    {
+      cleanup_free char *v_start = NULL;
+      char *end;
+
+      it = v_start = xstrdup (value);
+
+      end = strchr (it + 1, '\'');
+      if (end == NULL)
+        return crun_make_error (err, 0, "invalid variant `%s`", value);
+      *end = '\0';
+
+      sd_err = sd_bus_message_append (m, "(sv)", name, "s", it + 1);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus message append `%s`", name);
+    }
+  else if (has_prefix (it, "uint64"))
+    {
+      uint64_t v;
+
+      errno = 0;
+      v = strtoull (it + sizeof ("uint64"), NULL, 10);
+      if (UNLIKELY (errno != 0))
+        return crun_make_error (err, errno, "invalid value for `%s`", name);
+
+      sd_err = sd_bus_message_append (m, "(sv)", name, "t", v);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus message append `%s`", name);
+    }
+  else if (has_prefix (it, "int64"))
+    {
+      int64_t v;
+
+      errno = 0;
+      v = strtoll (it + sizeof ("int64"), NULL, 10);
+      if (UNLIKELY (errno != 0))
+        return crun_make_error (err, errno, "invalid value for `%s`", name);
+
+      sd_err = sd_bus_message_append (m, "(sv)", name, "x", v);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus message append `%s`", name);
+    }
+  else if (has_prefix (it, "uint32"))
+    {
+      uint32_t v;
+
+      errno = 0;
+      v = strtoul (it + sizeof ("uint32"), NULL, 10);
+      if (UNLIKELY (errno != 0))
+        return crun_make_error (err, errno, "invalid value for `%s`", name);
+
+      sd_err = sd_bus_message_append (m, "(sv)", name, "u", v);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus message append `%s`", name);
+    }
+  else if (has_prefix (it, "int32"))
+    {
+      int32_t v;
+
+      errno = 0;
+      v = strtol (it + sizeof ("int32"), NULL, 10);
+      if (UNLIKELY (errno != 0))
+        return crun_make_error (err, errno, "invalid value for `%s`", name);
+
+      sd_err = sd_bus_message_append (m, "(sv)", name, "i", v);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus message append `%s`", name);
+    }
+  else if ((strcmp (it, "true") == 0) || (strcmp (it, "false") == 0))
+    {
+      bool b = *it == 't';
+
+      sd_err = sd_bus_message_append (m, "(sv)", name, "b", b);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus message append `%s`", name);
+    }
+
```

Comment:
I'm sure you don't want to reimplement the entirety of gvariant parser here, yet a few notes:

1 gvariant assumes `int32` for value without an explicit type that can be parsed as a number. It might make sense to support it here.

2. I'd add checking of `endptr` from `strtoXX()` functions for being NULL (making sure there's no junk after the number).

3. There should be a whitespace after the type, i.e. `uint64123` is not valid but it is accepted by this parser.

At the very least I'd implement item 1.